### PR TITLE
New syntax and semantics of the @_specialize attribute

### PR DESCRIPTION
**Pull request by swiftix**
This PR implements the syntax and semantics of the new @_specialize attribute.

One can use the `@_specialize` attribute to force creation of specializations of a given function for a given set of generic parameters.

The new `@_specialize` attribute has a syntax like this:

```swift
@_specialize(exported: true, kind: full, where K == Int, V == Int)
@_specialize(exported: false, kind: partial, where K: _Trivial64)
func dictFunction<K, V>(dict: Dictionary<K, V>) {
}
```

If `exported` is set, the corresponding specialization would have a public visibility and can be used by clients.
If `exported` is omitted, it's value is assumed to be `false`.

If `kind` is `full` it means a full specialization.
If `kind` is `partial` it means a partial specialization.
If `kind` is omitted, its value is assumed to be `full`.

One can use in the `where` clause of the `@_specialize` attribute the newly introduced layout constraints like _Trivial, _Trivial(N) or _RefCountedObject, etc.

Partial specializations are supported syntactically, but their SIL level support is not committed as part of this PR. It will be provided as a separate PR.

### CODE REVIEW

**Comment by swiftix** [comment]
@slavapestov This is the second chunk of the #6486. It contains only the @_specialize related changes. No significant code changes compared to the parent PR. Could you please review?

**Comment by swiftix** [comment]
@swift-ci Please smoke test

File: include/swift/AST/Attr.h [hunk 1]
```diff
@@ -1024,34 +1027,84 @@ class SynthesizedProtocolAttr : public DeclAttribute {
...
+public:
+  enum class SpecializationKind {
+    Full,
+    Partial
+  };
 
-  TypeLoc *getTypeLocData() {
-    return reinterpret_cast<TypeLoc *>(this + 1);
+private:
+  AbstractFunctionDecl *F;
```
Comment (slavapestov) [inline include/swift/AST/Attr.h]: I think you shouldn't store the parent function inside the attribute, and just pass it in when necessary - what do you think?

File: lib/Parse/ParseGeneric.cpp [hunk 2]
```diff
@@ -261,22 +262,46 @@ ParserStatus Parser::parseGenericWhereClause(
...
+      if (Tok.is(tok::identifier) &&
+          getLayoutConstraintInfo(TypeId = Context.getIdentifier(Tok.getText()),
+                                  Context)) {
+        // Parse a layout constraint.
+        auto LayoutName = TypeId;
+        auto LayoutLoc = consumeToken();
+        auto LayoutInfo = parseLayoutConstraint(LayoutName);
+        if (!LayoutInfo.isKnownLayout()) {
+          // There was a bug in the layout constraint.
+          Status.setIsParseError();
```
Comment (slavapestov) [inline lib/Parse/ParseGeneric.cpp]: Is this an expected case or should it be an assertion?

File: lib/Parse/ParseGeneric.cpp [hunk 3]
```diff
@@ -261,22 +262,46 @@ ParserStatus Parser::parseGenericWhereClause(
...
-      
-      if (Protocol.isNull()) {
-        Status.setIsParseError();
-        if (Protocol.hasCodeCompletion())
-          Status.setHasCodeCompletion();
-        break;
+      Identifier TypeId;
+
+      if (Tok.is(tok::identifier) &&
+          getLayoutConstraintInfo(TypeId = Context.getIdentifier(Tok.getText()),
```
Comment (slavapestov) [inline lib/Parse/ParseGeneric.cpp]: This would be more readable if it was split into separate expressions in my opinion

File: include/swift/SILOptimizer/Utils/Generics.h [hunk 4]
```diff
@@ -52,20 +52,78 @@ class ReabstractionInfo {
...
+  /// The function type after applying the substitutions used to call the
+  /// specialized function.
   CanSILFunctionType SubstitutedType;
 
   /// The function type after applying the re-abstractions on the
   /// SubstitutedType.
   CanSILFunctionType SpecializedType;
 
+  /// The generic environment to be used by the specialization.
+  GenericEnvironment *SpecializedGenericEnv;
```
Comment (slavapestov) [inline include/swift/SILOptimizer/Utils/Generics.h]: Can this change go in the next PR instead?

File: lib/SILOptimizer/IPO/EagerSpecializer.cpp [hunk 5]
```diff
@@ -360,11 +437,149 @@ emitTypeCheck(SILBasicBlock *FailedTypeCheckBB, SubstitutableType *ParamTy,
...
+  auto GenericMT = Builder.createMetatype(
+    Loc, getThickMetatypeType(ContextTy->getCanonicalType()));
+
+  auto Int8Ty = SILType::getBuiltinIntegerType(8, Ctx);
+  auto BoolTy = SILType::getBuiltinIntegerType(1, Ctx);
+  Substitution Sub(ContextTy, {});
+
+  // Emit a check that it is a reference-counted object.
+  // TODO: Perform this check before all fixed size checks.
+  // FIXME: What builtin do we use to check it????
```
Comment (slavapestov) [inline lib/SILOptimizer/IPO/EagerSpecializer.cpp]: @jckarter might have a suggestion here

File: lib/SILOptimizer/IPO/EagerSpecializer.cpp [hunk 6]
```diff
@@ -385,17 +600,38 @@ SILValue EagerDispatch::emitArgumentCast(SILFunctionArgument *OrigArg,
...
 
-    auto CastArg = emitArgumentCast(OrigArg, Idx);
+    //auto CastArg = emitArgumentCast(SpecializedType, OrigArg, Idx);
+    auto CastArg = emitArgumentCast(SubstitutedType, OrigArg, Idx);
     DEBUG(dbgs() << "  Cast generic arg: "; CastArg->print(dbgs()));
 
-    if (ReInfo.isArgConverted(OrigArg->getIndex())) {
+    if (ReInfo.isArgConverted(OrigArg->getIndex()) &&
+        !isa<ArchetypeType>(
+            CastArg->getType().getSwiftRValueType()->getCanonicalType())) {
```
Comment (slavapestov) [inline lib/SILOptimizer/IPO/EagerSpecializer.cpp]: GetSwiftRValueType returns a CanType

File: lib/SILOptimizer/IPO/EagerSpecializer.cpp [hunk 7]
```diff
@@ -404,6 +640,15 @@ emitArgumentConversion(SmallVectorImpl<SILValue> &CallArgs) {
...
         // address we pass the loaded value.
+        // FIXME: If type of CastArg is an archetype, but it is loadable because
+        // of a layout constraint on the caller side, we have a problem here 
+        // We need to load the value on the caller side, but this archetype is
+        // not statically known to be loadable on the caller side (though we
+        // have proven dynamically that it has a fixed size).
+        // We can try to load it as an int value of width N, but then it is not
+        // clear how to convert it into a value of the archetype type, which is
+        // expected. May be we should pass it as @in parameter and make it
+        // loadable on the caller's side?
```
Comment (slavapestov) [inline lib/SILOptimizer/IPO/EagerSpecializer.cpp]: Yes, I think it has to be an @in parameter in this case

File: lib/SILOptimizer/IPO/EagerSpecializer.cpp [hunk 8]
```diff
@@ -448,7 +693,8 @@ static SILFunction *eagerSpecialize(SILFunction *GenericFunc,
   
   // Create a specialized function.
   GenericFuncSpecializer
-        FuncSpecializer(GenericFunc, SA.getSubstitutions(),
+        //FuncSpecializer(GenericFunc, SA.getSubstitutions(),
```
Comment (slavapestov) [inline lib/SILOptimizer/IPO/EagerSpecializer.cpp]: Remove the commented out line

File: lib/SILOptimizer/IPO/EagerSpecializer.cpp [hunk 9]
```diff
@@ -462,39 +708,54 @@ static SILFunction *eagerSpecialize(SILFunction *GenericFunc,
...
-    SmallVector<SILFunction*, 8> SpecializedFuncs;
+    SmallVector<SILFunction *, 8> SpecializedFuncs;
     SmallVector<ReabstractionInfo, 4> ReInfoVec;
     ReInfoVec.reserve(F.getSpecializeAttrs().size());
 
+    // TODO: Use a decision-tree to reduce the amount of dynamic checks being
+    // performed.
     for (auto *SA : F.getSpecializeAttrs()) {
-      ReInfoVec.emplace_back(&F, SA->getSubstitutions());
+      // ReInfoVec.emplace_back(ApplySite(), &F, SA->getSubstitutions());
```
Comment (slavapestov) [inline lib/SILOptimizer/IPO/EagerSpecializer.cpp]: Ditto

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 10]
```diff
@@ -122,6 +130,100 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
...
+CanSILFunctionType
+ReabstractionInfo::createSubstitutedType(SILFunction *OrigF,
+                                         const SubstitutionMap &SubstMap,
+                                         bool HasUnboundGenericParams) {
+  auto &M = OrigF->getModule();
+  auto OrigFnTy = OrigF->getLoweredFunctionType();
+
+  // First substitute concrete types into the existing function type.
+  auto FnTy = OrigFnTy->substGenericArgs(
+      M, QueryTypeSubstitutionMap{SubstMap.getMap()},
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Isn't there a version of substGenericArgs that takes a SubstitutionMap?

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 11]
```diff
@@ -171,6 +273,198 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
                          M.getASTContext());
 }
 
+std::pair<GenericEnvironment *, GenericSignature *>
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Can you leave this out for now and we will fix the issue it is working around

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 12]
```diff
@@ -171,6 +273,198 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
...
+  auto *GenericSig = Builder.getGenericSignature();
+  // Remember the new generic environment.
+  auto *GenericEnv = Builder.getGenericEnvironment(GenericSig);
+
+  return std::make_pair(GenericEnv, GenericSig);
+}
+
+/// Perform some sanity checks for the requirements
+static void
+checkSpecializationRequirements(ArrayRef<RequirementRepr> Requirements) {
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: I don't like seeing RequirementRepr used in SIL. They're not serialized for example

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 13]
```diff
@@ -171,6 +273,198 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
...
+      continue;
+    }
+
+    // TODO: Extend it to support TypeConstraint requirements once it is
+    // supported?
+    llvm_unreachable("Unknown type of requirement in generic specialization");
+  }
+}
+
+/// Remap Requirements to RequirementReps.
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Why is this needed?

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 14]
```diff
@@ -171,6 +273,198 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
...
+      OrigGenericSig->getSubstitutionMap(ForwardingSubs);
+  // Map archetypes of the original function to the contextual types
+  // of the specialized function.
+  SubstitutionMap ClonerArchetypeToConcreteMap;
+  SubstitutionMap CallerArchetypeToConcreteMap;
+
+  SmallVector<RequirementRepr, 4> RequirementReprs;
+  convertRequirements(Requirements, RequirementReprs);
+
+  for (auto &Req : Requirements) {
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: You can just iterate over the original requirements here no?

File: lib/Serialization/Deserialization.cpp [hunk 15]
```diff
@@ -3408,9 +3461,16 @@ Decl *ModuleFile::getDecl(DeclID DID, Optional<DeclContext *> ForcedContext) {
...
   // Record the attributes.
-  if (DAttrs)
+  if (DAttrs) {
     declOrOffset.get()->getAttrs().setRawAttributeChain(DAttrs);
-
+    // @_specialize attributes need to know their decl context.
+    for (auto Attr : declOrOffset.get()->getAttrs()) {
+      if (Attr->getKind() != DAK_Specialize)
+        continue;
+      reinterpret_cast<SpecializeAttr *>(Attr)->setDecl(
```
Comment (slavapestov) [inline lib/Serialization/Deserialization.cpp]: Do Attrs not support llvm casts?

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 16]
```diff
@@ -171,6 +273,198 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
...
+
+  // Substitute into forwarding substitutions to get the cloner substitutions
+  // and substitutions to be used by the caller for calling the specialized
+  // function.
+  SmallVector<Substitution, 4> ClonerSubs;
+  SmallVector<Substitution, 4> CallerSubs;
+  for (auto Sub : ForwardingSubs) {
+    auto NewClonerSub = Sub.subst(
+        SM, QueryTypeSubstitutionMap{ClonerArchetypeToConcreteMap.getMap()},
+        LookUpConformanceInModule(SM));
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: I'd like us to figure out why this can't find the conformances in the SubstitutionMap before we merge this part -- is that OK?

**Comment by slavapestov** [comment]
I see the SIL changes are in later commits. Do the earlier commits for the parser change compile and pass tests already?

File: lib/Parse/ParseGeneric.cpp (hunk 2 above)
Comment (swiftix) [inline lib/Parse/ParseGeneric.cpp]: It is expected. The input may be syntactically wrong and which case we would get the UnknownLayoutConstraint back.

**Comment by swiftix** [comment]
@slavapestov 

> I see the SIL changes are in later commits. Do the earlier commits for the parser change compile and pass tests already?

I'm not sure what you are asking about. The tests in the last commit cover both Swift and SIL parsing of the @_specialized attribute.

File: include/swift/AST/Attr.h (hunk 1 above)
Comment (swiftix) [inline include/swift/AST/Attr.h]: Yes. In general I didn't want to store it in the attribute. But in some contexts (e.g. in the printer) we currently don't pass the FuncDecl around. I'll see if it can done without too many changes.

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 16 above)
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: Sure, I'd love to.  I even have a small test-case demonstrating the issue. We can discuss it in person or in a chat.

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 13 above)
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: Good catch. I've written this as a workaround a while ago and it is most likely outdated. IIRC, there were some issues with the ArchetypeBuilder APIs at that time. Somehow, only RequirementReprs could be added to the builder using the APIs or something like this. But we don't need this hack anymore. I'll remove it if everything builds OK without it.

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 11 above)
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: What exactly should be left out? I do not quite follow.

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 13 above)
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: Yes. I was able to remove all occurrences of RequirementRepr from Generics.cpp and everything seems to work just fine.

File: lib/Serialization/Deserialization.cpp (hunk 15 above)
Comment (swiftix) [inline lib/Serialization/Deserialization.cpp]: Fixed.

File: lib/Parse/ParseGeneric.cpp (hunk 3 above)
Comment (swiftix) [inline lib/Parse/ParseGeneric.cpp]: Fixed.

File: include/swift/SILOptimizer/Utils/Generics.h (hunk 4 above)
Comment (swiftix) [inline include/swift/SILOptimizer/Utils/Generics.h]: Do you mean the SpecializedGenericEnv change specifically or almost all of the Generics.h API changes?

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 10 above)
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: Yeah, I think it used to exist. But it was removed recently.

File: include/swift/SILOptimizer/Utils/Generics.h (hunk 4 above)
Comment (slavapestov) [inline include/swift/SILOptimizer/Utils/Generics.h]: The latter. It looks like it's a separate patch anyway.

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 11 above)
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: At some point you had a utility function that fed a generic signature back through an archetype builder to get a 'more canonical' signature. This was a bug since it should already be canonical the first time. But I see this function does something else, so never mind!

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 16 above)
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Great. I wonder if it's related to https://bugs.swift.org/browse/SR-3500. The reason I wanted to split this PR further is so we can get the parser changes checked in before we figure out this part.

**Comment by slavapestov** [comment]
> I'm not sure what you are asking about. The tests in the last commit cover both Swift and SIL parsing of the @_specialized attribute.

Sure, but the earlier patches that only touch the parser could go in independently of the later changes right? Just trying to get as much of this checked in as quickly as possible :-)

File: include/swift/AST/DiagnosticsParse.def [hunk 17]
```diff
@@ -702,7 +702,7 @@ ERROR(layout_alignment_should_be_positive,none,
 ERROR(expected_rparen_layout_constraint,none,
       "expected ')' to complete layout constraint", ())
 ERROR(layout_constraints_only_inside_specialize_attr,none,
-      "layout constraints are only allowed inside @_specialize attributes", ())
+      "layout constraints are only allowed inside '_specialize' attributes", ())
```
Comment (slavapestov) [inline include/swift/AST/DiagnosticsParse.def]: Hmm... I think the other diagnostics (eg the ones for @objc) still have the '@' sign in the message

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 10 above)
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: That's odd. Really the `ArrayRef<Substitution>` version should be removed in favor of a SubstitutionMap one. I think it should be added back. :-)

**Comment by slavapestov** [comment]
Just looked over the latest patches. The parser/serialization/AST bits LGTM now.

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 16 above)
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: I looked at the referenced SR and indeed the stack trace looks very similar to what I was observing.

File: include/swift/SILOptimizer/Utils/Generics.h (hunk 4 above)
Comment (swiftix) [inline include/swift/SILOptimizer/Utils/Generics.h]: I see. I'll try to move SIL generation changes into a subsequent patch. Eventually, I'll have to disable the EagerSpecializer temporarily, because it needs to use the new @_specialize attribute for the codegen, but won't be able to do it without Generics.cpp changes.

**Comment by swiftix** [comment]
@slavapestov I removed all Generics.cpp and related changes. The PR contains now only parsing, type checks and tests. Eager specializer and related tests are disabled for now.

**Comment by swiftix** [comment]
@swift-ci Please smoke test

**Comment by swiftix** [comment]
@swift-ci Please smoke test

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 16 above)
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: @slavapestov I've sent you an email with the detailed description of the issue.

File: include/swift/AST/LayoutConstraint.h [hunk 18]
```diff
@@ -48,7 +48,7 @@ enum class LayoutConstraintKind : unsigned char {
 /// This is a class representing the layout constraint.
 class LayoutConstraintInfo {
   // Alignment of the layout in bytes.
-  unsigned Alignment;
+  unsigned Alignment : 32;
```
Comment (slavapestov) [inline include/swift/AST/LayoutConstraint.h]: You should only need < 16 bits for alignment

File: include/swift/AST/Attr.h [hunk 19]
```diff
@@ -1024,30 +1029,71 @@ class SynthesizedProtocolAttr : public DeclAttribute {
   }
 };
 
+class AbstractFunctionDecl;
```
Comment (slavapestov) [inline include/swift/AST/Attr.h]: Forward declarations should be at the top of the file

File: include/swift/AST/LayoutConstraint.h (hunk 18 above)
Comment (slavapestov) [inline include/swift/AST/LayoutConstraint.h]: You should only need < 16 bits for alignment

**Comment by slavapestov** [comment]
This looks great. I added a couple of minor nit-picks but feel free to address them in a follow-up PR.

**Comment by swiftix** [comment]
@slavapestov Thanks for reviewing the new version of the PR. I'll add these last two fixes you suggested. 

BTW, any news on the conformance lookup issue? Have you looked at the example I sent you? Is it a bug in the conformance lookup implementation? Or may be the SubstitutionMap was created in a wrong way?

**Comment by slavapestov** [comment]
I saw your note about the conformance lookup issue but I haven't investigated it yet. Thanks for the detailed explanation though!

**Comment by swiftix** [comment]
@swift-ci Please smoke test

**Comment by swiftix** [comment]
@swift-ci Please smoke test Linux
